### ISSUE TITLE
KPI of allenai/OLMo-2-0425-1B-SFT is not reproducible

### ISSUE THREAD

**Issue by Yi-Yu-Yvonne** [allenai/open-instruct#751]
I am attempting to reproduce the evaluation scores reported on Hugging Face for the model [`allenai/OLMo-2-0425-1B-SFT`](https://huggingface.co/allenai/OLMo-2-0425-1B-SFT), but I am seeing inconsistent results.

This seems similar to [Issue #688](https://github.com/allenai/open-instruct/issues/688), but that issue is still unresolved.

I used the following commands to evaluate the model:
```bash
olmes --model allenai/OLMo-2-0425-1B-SFT --task ifeval::tulu --output-dir "output/ifeval"
olmes --model allenai/OLMo-2-0425-1B-SFT --task gsm8k::tulu --output-dir "output/gsm8k"
```

## My results
- `ifeval::tulu`: **0.480591**
- `gsm8k::tulu`: **0.518575**

## Expected results (from [Hugging Face model card](https://huggingface.co/allenai/OLMo-2-0425-1B-SFT))
- `ifeval::tulu`: **50.5**
- `gsm8k::tulu`: **52.1**

Thanks in advance!

**Comment by hamishivi** [comment]
Hi, thanks for your interest! I think the point or so difference is likely due to differences in package versions (e.g. vllm). For example, changing vllm version can yield a 1-2 point change like you see here. The exact commands used to run eval were:

```bash
# ifeval
python -m oe_eval.run_eval --task '{"task_name": "ifeval", "primary_metric": "prompt_level_loose_acc", "split": "train", "use_chat_format": true, "generation_kwargs": {"max_gen_toks": 4096, "do_sample": false, "temperature": 0.0, "stop_sequences": [], "truncate_context": false}, "metric_kwargs": {"aggregation_levels": ["prompt", "inst"], "strictness_levels": ["strict", "loose"], "output_individual_metrics": true}, "metadata": {"regimes": ["Tulu"], "alias": "ifeval::tulu"}}' --batch-size 10000 --save-raw-requests true --num-workers 1 --gpus 1 --model olmo2_1b_v2_sft_lr3e-5_seed1__1__1744989064 --model-args '{"model_path": "gs://ai2-llm/post-training//nathanl/output/olmo2_1b_v2_sft_lr3e-5_seed1__1__1744989064", "max_length": 4096, "model_type": "vllm"}' --push-datalake --datalake-tags run_id=https://wandb.ai/ai2-llm/open_instruct_internal/runs/532v35jn --output-dir /results
# gsm8k 
python -m oe_eval.run_eval --task '{"task_name": "gsm8k", "split": "test", "primary_metric": "exact_match", "use_chat_format": true, "num_shots": 8, "fewshot_source": "STD:GSM8k", "random_subsample_seed": 42, "chat_overrides": {"context_kwargs": {"fewshot_as_multiturn": true}}, "metadata": {"regimes": ["Tulu"], "alias": "gsm8k::tulu"}, "generation_kwargs": {"max_gen_toks": 4096, "truncate_context": false}}' --batch-size 10000 --save-raw-requests true --num-workers 1 --gpus 1 --model olmo2_1b_v2_sft_lr3e-5_seed1__1__1744989064 --model-args '{"model_path": "gs://ai2-llm/post-training//nathanl/output/olmo2_1b_v2_sft_lr3e-5_seed1__1__1744989064", "max_length": 4096, "model_type": "vllm"}' --push-datalake --datalake-tags run_id=https://wandb.ai/ai2-llm/open_instruct_internal/runs/532v35jn --output-dir /results
```
I think these should match up with your commands (they are just the expanded form), and likely package version differences is why there is some difference.

As for training, we used:
```
accelerate launch --mixed_precision bf16 --num_processes 8 --use_deepspeed --deepspeed_config_file configs/ds_configs/stage3_no_offloading_accelerate.conf --deepspeed_multinode_launcher standard open_instruct/finetune.py --exp_name olmo2_1b_v2_sft_lr3e-5_seed1 --model_name_or_path allenai/OLMo-2-0425-1B --model_revision main --tokenizer_name allenai/OLMo-2-1124-7B --tokenizer_revision main --use_slow_tokenizer False --add_bos --dataset_mixer_list allenai/tulu-3-sft-olmo-2-mixture-0225 1.0 --use_flash_attn --max_seq_length 4096 --per_device_train_batch_size 2 --gradient_accumulation_steps 8 --learning_rate 3e-5 --lr_scheduler_type linear --warmup_ratio 0.03 --weight_decay 0.0 --num_train_epochs 2 --reduce_loss sum --report_to wandb --with_tracking --logging_steps 1 --seed 1
```
Training was run on one node. I think the commands look the same minus the one node/two node difference? Not sure what else could explain differences apart from again package changes....

**Comment by yi-yvonne-yu** [comment]
Hi, thanks for the response!

I was wondering if the evaluation results shown on the [Hugging Face model card](https://huggingface.co/allenai/OLMo-2-0425-1B-SFT) were generated using a different package setup than what's provided in https://github.com/allenai/olmes?
I set up the environment by following the installation instructions in olmes and ran the evaluation using the same commands (just replacing the --model name, model_path and removing --push-datalake --datalake-tags), but I’m still seeing a small performance gap.

My results:
- ifeval::tulu: 0.5009
- gsm8k::tulu: 0.5163

Expected results (from the [model card](https://huggingface.co/allenai/OLMo-2-0425-1B-SFT)):
- ifeval::tulu: 50.5
- gsm8k::tulu: 52.1

---

I also tried training with: 1 node, 4 GPUs, 4 processes, 
gradient_accumulation_steps=16. All other parameters (e.g., learning rate, max sequence length, epochs, dataset, etc.) were kept the same.

The effective batch size remains the same as the 8-GPU setup, but the evaluation results are noticeably different and lower than expected. Do you have any idea what might be causing this difference? I’m using the setup as indicated in the README.md.

Thank you!

**Comment by hamishivi** [comment]
Yeah, the evaluation results were run using OLMES (or rather, an internal fork of it), on a static image we have. OLMES doesn't have tight pinning of package versions, so that can result in changes. For example, I changed vllm version and observed a ~1 point difference in GSM8k for this model. I think within 1 point is about as good as we can get without pinning all the software versions to be the exact same. I've attached the exact predictions from our old run of OLMES that we copied the scores from to the HF card, maybe thats useful? 

As for training, the command seems right.... do you have the exact eval scores for these different runs? I'm not entirely sure what could cause these changes, or if the 1B model has some larger variance across seeds during training.

[task-000-gsm8k-predictions (11).jsonl.zip](https://github.com/user-attachments/files/21132774/task-000-gsm8k-predictions.11.jsonl.zip)

**Comment by yi-yvonne-yu** [comment]
Thanks! One thing I noticed — the training script uses --tokenizer_name allenai/OLMo-2-1124-7B, but in finetune.py, the tokenizer seems to automatically resolve to allenai/OLMo-2-0425-1B. Could this mismatch be contributing to the inconsistent results?

**Comment by hamishivi** [comment]
Sorry, i dropped the ball on this one - the tokenizer shouldn't matter, since its the same across olmo 2 models.

**Comment by finbarrtimbers** [comment]
Closing this due to age; feel free to reopen if you have additional questions.
